lib/rpc/src/tcp_client.rs

```diff
@@ -0,0 +1,463 @@
+            socket
+                .connect((self.server_ip, self.server_port), self.client_port)
+                .unwrap();
+            debug!(
+                "Attempting to connect to server {}:{}",
+                self.server_ip, self.server_port
+            );
+        }
+
+        // Connect to server
+        loop {
+            match self.iface.poll(&mut self.sockets, Instant::from_millis(0)) {
+                Ok(_) => {}
+                Err(e) => {
+                    warn!("poll error: {}", e);
+                }
+            }
+            let socket = self.sockets.get::<TcpSocket>(self.server_handle.unwrap());
+            // Waiting for send/recv forces the TCP handshake to fully complete
+            if socket.is_active() && (socket.may_send() || socket.may_recv()) {
+                debug!("Connected to server, ready to send/recv data");
+                break;
+            }
+        }
+
+        self.rpc_call(0, RPCType::Registration, Vec::new()).unwrap();
+        Ok(self.client_id)
+    }
+}
+
+/// RPC client operations
+impl RPCClientAPI for TCPClient<'_> {
+    /// calls a remote RPC function with ID
+    fn rpc_call(
+        &mut self,
+        pid: usize,
+        rpc_id: RPCType,
+        data: Vec<u8>,
+    ) -> Result<Vec<u8>, RPCError> {
+        // Create request header
+        let req_hdr = RPCHeader {
+            client_id: self.client_id,
+            pid: pid,
+            req_id: self.req_id,
+            msg_type: rpc_id,
+            msg_len: data.len() as u64,
+        };
+
+        // Serialize request header then request body
+        let mut req_data = Vec::new();
+        unsafe { encode(&req_hdr, &mut req_data) }.unwrap();
+        if data.len() > 0 {
+            //unsafe { encode(&data, &mut req_data) }.unwrap();
+            req_data.extend(data);
+        }
+
+        // Send request
+        self.msg_send(req_data).unwrap();
+
+        // Receive response header
+        let mut res_data = self.msg_recv(core::mem::size_of::<RPCHeader>()).unwrap();
+        let (res_hdr, extra) = unsafe { decode::<RPCHeader>(&mut res_data) }.unwrap();
+        assert_eq!(extra.len(), 0);
+
+        // Read the rest of the data
+        let mut payload_data = Vec::new();
+        if res_hdr.msg_len > 0 {
+            payload_data = self.msg_recv(res_hdr.msg_len as usize).unwrap();
+        }
+
+        // Check request & client IDs, and also length of received data
+        if ((res_hdr.client_id != self.client_id) && rpc_id != RPCType::Registration)
+            || res_hdr.req_id != self.req_id
+        {
+            warn!(
+                "Mismatched client id ({}, {}) or request id ({}, {})",
+                res_hdr.client_id, self.client_id, res_hdr.req_id, self.req_id
+            );
+            return Err(RPCError::MalformedResponse);
+        }
+
+        // Increment request id
+        self.req_id += 1;
+
+        // If registration, update id
+        if rpc_id == RPCType::Registration {
```

Comment:
The special handling for RPCType::Registration might fit better if moved into the join cluster function?
ideally we should think if there is a way with the type system to prevent sending a message without registration (aka have a builderpattern for RPCClientAPi that registers first and then returns a type with initialized client_id? Or ensure register message is sent in the constructor and then can't be ever sent again?